src/rust/engine/process_execution/src/remote_cache.rs

```diff
@@ -0,0 +1,458 @@
+use std::collections::{BTreeMap, VecDeque};
+use std::convert::TryInto;
+use std::path::Component;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use bazel_protos::call_option;
+use bazel_protos::remote_execution::{
+  ActionResult, Command, FileNode, Tree, UpdateActionResultRequest,
+};
+use fs::RelativePath;
+use futures::compat::Future01CompatExt;
+use hashing::Digest;
+use store::Store;
+use workunit_store::{with_workunit, Level, WorkunitMetadata};
+
+use crate::remote::make_execute_request;
+use crate::{
+  Context, FallibleProcessResultWithPlatform, MultiPlatformProcess, Platform, Process,
+  ProcessMetadata,
+};
+
+/// This `CommandRunner` implementation caches results remotely using the Action Cache service
+/// of the Remote Execution API.
+///
+/// This runner expects to sit between the local cache CommandRunner and the CommandRunner
+/// that is actually executing the Process. Thus, the local cache will be checked first,
+/// then the remote cache, and then execution (local or remote) as necessary if neither cache
+/// has a hit. On the way back out of the stack, the result will be stored remotely and
+/// then locally.
```

Comment:
Thanks for the excellent comment! I'm a bit unclear how the execution `CommandRunner` is different than the local cache `CommandRunner`? I'm not very familiar with this part of the codebase.

Reply:
Do you mean the remote execution command runner (`process_execution::remote::CommandRunner`) or the remote cache command runner (`process_execution::remote_cache::CommandRunner`)?

Reply:
I don't understand what this sentence means:

> This runner expects to sit between the local cache CommandRunner and the CommandRunner that is actually executing the Process.

Reply:
It means that the nested "stack" of command runners looks like this:

`local_cache::ComandRunner` -> `remote_cache::CommandRunner` -> `local::ComandRunner`

The actual execution of the Process is done by `local::CommandRunner` (the one on the right).  But before it is invoked, we first check the local cache and then check the remote cache. On the way back up the stack, the `remote_cache::CommandRunner` will take the result of any actual invocation performed by `local::ComandRunner` and write it to the remote Action Cache.